### PR TITLE
* Improved support for MuseScore 3.0 plugins (Supersedes PR #3297)

### PR DESCRIPTION
**Pull request by rjmatthews62**
* Implemented TimeSig wrapper so random.qml would work.
* Added a "set" to ElementW for setting properties.
* Added a buildWrapper static function to allow different wrapper descendants.
* Converted "random.qml" to work with MuseScore 3.0
* Added a TimeSigW class to support added features required to manipulate TimeSig
* Modified timesig help
* Added support for lists of valid property names for use with "get" and "set"
* colornotes.qml now works with Musescore 3.0
* Added proplist.qml to demonstrate readProperties and writeProperties
* Modified help system to recognize wrappers.
* Moved ElementW into element.h and element.cpp
* Started updating help annotations.
* Plugin support for lyrics. Improved help files.
* Fix after applying to wrong branch.
* notenames.qml now works.

### CODE REVIEW

File: libmscore/chord.h [hunk 1]
```diff
@@ -58,9 +63,32 @@ enum class PlayEventType : char    {
 //   @P stemDirection Direction       the stem slash of the chord (acciaccatura) if any (read only)
 //---------------------------------------------------------
 
+class ChordW : public ChordRestW {
+      Q_OBJECT
```
Comment (Jojo-Schmitz) [inline libmscore/chord.h]: wouldn't `Q_GADGET` be enough here?

File: libmscore/chordrest.h [hunk 2]
```diff
@@ -40,13 +40,35 @@ class Spanner;
...
+//   @W ChordRest
 ///    Virtual base class. Chords and rests can be part of a beam
-//
+///   <br>named properties (get/set): <i>small, beam_mode, staff_move, duration_type</i>
 //   @P beamMode      enum (Beam.AUTO, .BEGIN, .MID, .END, .NONE, .BEGIN32, .BEGIN64, .INVALID)
 //   @P durationType  int
 //   @P small         bool           small chord/rest
 //-------------------------------------------------------------------
+class ChordRestW : public DurationElementW {
+      Q_OBJECT
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.h]: wouldn't `Q_GADGET` be enough here?

Comment (rjmatthews62) [reply libmscore/chordrest.h]: Actually, no, not as far as I can tell. Has to be a descendant of QObject to export.
Werner felt so too, if you look at the definitions for what was exposed in the current master branch. they are all QObjects (with a Q_OBJECT annotation)

Comment (Jojo-Schmitz) [reply libmscore/chordrest.h]: OK, I was just wondering...

Comment (rjmatthews62) [reply libmscore/chordrest.h]: Just double checked, and ALL the qml interface documentation refers to Q_OBJECT... I'm pretty sure it's why Werner started implementing wrappers in the first place. (I think... I don't pretend to be an expert)

Comment (Jojo-Schmitz) [reply libmscore/chordrest.h]: Or why he left those at Q_OBJECT and only changed the others to Q_GADGET

**Comment by Jojo-Schmitz** [comment]
See https://travis-ci.org/musescore/MuseScore/jobs/284043085#L4616-L4639, there is a change needed in mtest/libmscore/dynamic/tst_dynamic.cpp, from `Ms::Element::Placement` to `Ms::Placement`

**Comment by rjmatthews62** [comment]
Thanks... I was just hunting through Travis for the error, but it's being very slow ATM.
Is there an easy way to compile and run mtest under windows? I've managed to get bits of it to work, but not the test suite as a whole.

**Comment by Jojo-Schmitz** [comment]
I've never got mtest to (fully) work on Windows, so I let Travis do it for me ;-)

File: libmscore/duration.h [hunk 3]
```diff
@@ -24,14 +25,33 @@ class Beam;
...
+      Q_PROPERTY(FractionWrapper* globalDuration READ globalDurW)
+   public:
+      DurationElementW() : ElementW() {}
+      DurationElementW(ScoreElement* _e) : ElementW(_e) {}
+      DurationElement* durationElement();
+      FractionWrapper* durationW();
+      void setDurationW(FractionWrapper* f);
+      FractionWrapper* globalDurW();
+      };
+//@E
```
Comment (Jojo-Schmitz) [inline libmscore/duration.h]: ??

**Comment by Jojo-Schmitz** [comment]
I believe "fixes #257631 Crash in layout with TieSegments" belonqs into a separate PR

File: libmscore/chordrest.h [hunk 4]
```diff
@@ -40,13 +40,35 @@ class Spanner;
 enum class SegmentType;
 
 //-------------------------------------------------------------------
-//   @@ ChordRest
+//   @@ ChordRestW
+//   @W ChordRest
 ///    Virtual base class. Chords and rests can be part of a beam
-//
+///   <br>named properties (get/set): <i>small, beam_mode, staff_move, duration_type</i>
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.h]: Wouldn't this styling better be dealt with in the code that generates the plugin docs, rather then embedding HTML here?

Comment (rjmatthews62) [reply libmscore/chordrest.h]: Probably, but at present it's a work in progress.

**Comment by rjmatthews62** [comment]
fixes #257631 IS in a separate PR, but I needed to also include it in this branch so it would stop crashing on me.

**Comment by Jojo-Schmitz** [comment]
Ah, I see, it is (also) in #3308

File: libmscore/chordrest.h (hunk 2 above)
Comment (rjmatthews62) [reply libmscore/chordrest.h]: I did some experiments, and Q_GADGET does NOT play well with Qml. I wonder if it has something to do with 5.8 vs 5.9?

File: libmscore/chordrest.h (hunk 4 above)
Comment (rjmatthews62) [reply libmscore/chordrest.h]: Latest update now implements a @S annotation for this purpose.

File: libmscore/chordrest.cpp [hunk 5]
```diff
@@ -1109,6 +1109,16 @@ bool ChordRest::setProperty(P_ID propertyId, const QVariant& v)
...
       }
 
+void ChordRest::supportedProperties(QList<P_ID>& dest, bool writeable)
+      {
+      DurationElement::supportedProperties(dest, writeable);
+      if (writeable) {
+            dest << P_ID::SMALL << P_ID::BEAM_MODE << P_ID::STAFF_MOVE << P_ID::DURATION_TYPE << P_ID::VISIBLE;
+            } else {
+            dest << P_ID::SMALL << P_ID::BEAM_MODE << P_ID::STAFF_MOVE << P_ID::DURATION_TYPE;
+            }
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.cpp]: IMHO better with less braces here, also further up in other single line stamtents following an `if` or `for`.
Makes it easier to read, I think. 
Here not even an `else` seems needed, if the statement following were moved to before the `if`?
Matter if personal taste, mostly...

File: manual/genManual.cpp [hunk 6]
```diff
@@ -302,6 +323,9 @@ static void writeOutput()
                         }
                   out += "</table></div>\n";
                   }
+            if (!cl.properties.isEmpty()) {
+                  out += "\n<h4>Named properties (get/set)</h4>\n<i>"+cl.properties.replace(",",", ") +"</i>";
```
Comment (Jojo-Schmitz) [inline manual/genManual.cpp]: speces around the 2 `+`

**Comment by orchestralmusic-net** [comment]
@anatoly-os Did you forget to add label "Plugins" here?

**Comment by Jojo-Schmitz** [comment]
I don't think a label "Plugins" exists?

This PR needs a rebase, but it may well be broken beyond repair and may need a complete rewrite instead.. A lot has changed between October 2017 and now...

**Comment by orchestralmusic-net** [comment]
Well:
https://github.com/musescore/MuseScore/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3APlugins

**Comment by anatoly-os** [comment]
This PR is far behind the actual state of the Plugins API infrastructure

**Pull request closed without merging**